File: src/routes/policy.rs
//! OLP routing policy. Restrictions are cumulative; preferences never grant access.
use crate::ids::{ProviderId, RouteSlug};
use crate::providers::options::ConnectionOptions;
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use std::collections::{BTreeMap, BTreeSet};
use utoipa::ToSchema;

#[derive(
    Clone, Copy, Debug, Default, Deserialize, Serialize, ToSchema, Eq, PartialEq, Ord, PartialOrd,
)]
#[serde(rename_all = "snake_case")]
pub enum RoutingStrategy {
    #[default]
    Weighted,
    Price,
    Latency,
    Throughput,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize, ToSchema, Eq, PartialEq)]
#[serde(default, deny_unknown_fields)]
pub struct PriceCeiling {
    pub input_per_million: Option<String>,
    pub output_per_million: Option<String>,
    pub unit_price: Option<String>,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize, ToSchema, Eq, PartialEq)]
#[serde(default, deny_unknown_fields)]
pub struct RoutingConstraints {
    pub only: Option<Vec<String>>,
    pub ignore: Vec<String>,
    pub regions: Option<BTreeSet<String>>,
    pub quantizations: Option<BTreeSet<String>>,
    pub deny_data_collection: bool,
    pub require_zero_data_retention: bool,
    pub require_parameters: bool,
    pub max_price: Option<PriceCeiling>,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize, ToSchema, Eq, PartialEq)]
#[serde(default, deny_unknown_fields)]
pub struct RoutingPreferences {
    #[serde(skip)]
    #[schema(ignore)]
    pub cooling_slots: BTreeSet<uuid::Uuid>,
    #[serde(skip)]
    #[schema(ignore)]
    pub required_credential_version: Option<uuid::Uuid>,
    #[serde(flatten)]
    pub constraints: RoutingConstraints,
    pub order: Option<Vec<String>>,
    pub strategy: Option<RoutingStrategy>,
    pub allow_fallbacks: Option<bool>,
    pub preferred_max_latency_ms: Option<u64>,
    pub preferred_min_throughput: Option<u64>,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize, ToSchema, Eq, PartialEq)]
#[serde(default, deny_unknown_fields)]
pub struct RoutingPolicy {
    pub constraints: RoutingConstraints,
    pub defaults: RoutingPreferences,
    pub allowed_strategies: Option<BTreeSet<RoutingStrategy>>,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[serde(default)]
pub struct RoutingConfiguration {
    pub version: u16,
    pub installation: RoutingPolicy,
    pub routes: BTreeMap<RouteSlug, RoutingPolicy>,
    pub providers: BTreeMap<ProviderId, ConnectionOptions>,
    pub credentials: BTreeMap<ProviderId, Vec<crate::providers::pool::CredentialSlot>>,
    /// Current published restrictions also constrain retained releases. Secret
    /// versions and transport configuration remain pinned to their release.
    #[serde(skip)]
    pub credential_authority:
        Option<BTreeMap<ProviderId, Vec<crate::providers::pool::CredentialSlot>>>,
    /// Explicitly revoked credential versions. A retained release keeps its
    /// pinned version and transport, but selection refuses the revoked version.
    #[serde(skip)]
    pub revoked_credential_versions: BTreeSet<uuid::Uuid>,
    #[serde(skip)]
    pub connection_limit_authority:
        Option<BTreeMap<ProviderId, crate::providers::options::ConnectionLimits>>,
    pub prices: Vec<RoutingPrice>,
}

#[derive(Clone, Debug, Deserialize, Serialize, ToSchema)]
pub struct RoutingPrice {
    pub effective_at: chrono::DateTime<chrono::Utc>,
    pub scope_priority: u8,
    pub revision: i32,
    pub revision_id: uuid::Uuid,
    pub provider_id: uuid::Uuid,
    pub model: String,
    pub operation: String,
    pub currency: String,
    pub input_per_million: Option<String>,
    pub output_per_million: Option<String>,
    pub unit_price: Option<String>,
}

pub fn decimal(value: &str) -> Result<Decimal, String> {
    crate::usage::pricing::validate_decimal(value)
        .map_err(|_| "Invalid non-negative decimal price")?;
    value.parse().map_err(|_| "Price is out of range".into())
}

pub fn validate_selector(selector: &str) -> Result<(), String> {
    if let Some(id) = selector.strip_prefix("vendor:")
        && crate::providers::catalog::vendor(id).is_some()
    {
        return Ok(());
    }
    if let Some(id) = selector.strip_prefix("provider:")
        && uuid::Uuid::parse_str(id).is_ok()
    {
        return Ok(());
    }
    Err("Use vendor:<catalog-id> or provider:<uuid> selectors".into())
}

pub fn matches(selector: &str, provider: ProviderId, vendor: Option<&str>) -> bool {
    selector
        .strip_prefix("provider:")
        .and_then(|s| uuid::Uuid::parse_str(s).ok())
        == Some(provider.as_uuid())
        || selector
            .strip_prefix("vendor:")
            .is_some_and(|s| Some(s) == vendor)
}

impl RoutingConstraints {
    pub fn validate(&self) -> Result<(), String> {
        if self.only.as_ref().is_some_and(Vec::is_empty)
            || self.regions.as_ref().is_some_and(BTreeSet::is_empty)
            || self.quantizations.as_ref().is_some_and(BTreeSet::is_empty)
        {
            return Err("An allowlist must not be empty".into());
        }
        for selectors in [self.only.as_deref().unwrap_or_default(), &self.ignore] {
            if selectors.len() > 100 {
                return Err("At most 100 selectors are allowed".into());
            }
            for selector in selectors {
                validate_selector(selector)?;
            }
        }
        if let Some(ceiling) = &self.max_price {
            for value in [
                &ceiling.input_per_million,
                &ceiling.output_per_million,
                &ceiling.unit_price,
            ]
            .into_iter()
            .flatten()
            {
                decimal(value)?;
            }
        }
        Ok(())
    }
}

impl RoutingPreferences {
    pub fn validate(&self) -> Result<(), String> {
        self.constraints.validate()?;
        if let Some(order) = &self.order {
            if order.is_empty() || order.len() > 100 {
                return Err("Order requires between one and 100 selectors".into());
            }
            for selector in order {
                validate_selector(selector)?;
            }
        }
        if self.preferred_max_latency_ms == Some(0) || self.preferred_min_throughput == Some(0) {
            return Err("Performance preferences must be positive".into());
        }
        Ok(())
    }

    pub fn from_headers(headers: &http::HeaderMap) -> Result<Self, String> {
        let values = headers.get_all("x-olp-routing");
        if values.iter().count() > 1 {
            return Err("Supply X-OLP-Routing once".into());
        }
        let Some(value) = values.iter().next() else {
            return Ok(Self::default());
        };
        let value: Self = serde_json::from_slice(value.as_bytes())
            .map_err(|_| "X-OLP-Routing must contain a valid routing preferences object")?;
        value.validate()?;
        Ok(value)
    }

    pub fn overlay(&mut self, other: &Self) {
        if other.order.is_some() {
            self.order.clone_from(&other.order);
        }
        if other.strategy.is_some() {
            self.strategy = other.strategy;
        }
        if other.allow_fallbacks.is_some() {
            self.allow_fallbacks = other.allow_fallbacks;
        }
        if other.preferred_max_latency_ms.is_some() {
            self.preferred_max_latency_ms = other.preferred_max_latency_ms;
        }
        if other.preferred_min_throughput.is_some() {
            self.preferred_min_throughput = other.preferred_min_throughput;
        }
    }
}

impl RoutingPolicy {
    pub fn validate(&self) -> Result<(), String> {
        self.constraints.validate()?;
        self.defaults.validate()?;
        if self
            .allowed_strategies
            .as_ref()
            .is_some_and(BTreeSet::is_empty)
        {
            return Err("At least one routing strategy must be allowed".into());
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn rejects_ambiguous_headers_and_invalid_selectors() {
        let mut headers = http::HeaderMap::new();
        headers.insert(
            "x-olp-routing",
            r#"{"only":["vendor:deepseek"],"strategy":"price"}"#
                .parse()
                .unwrap(),
        );
        assert!(RoutingPreferences::from_headers(&headers).is_ok());
        headers.append("x-olp-routing", "{}".parse().unwrap());
        assert!(RoutingPreferences::from_headers(&headers).is_err());
        assert!(validate_selector("vendor:made-up").is_err());
        for value in [
            r#"{"max_attempts":999}"#,
            r#"{"required_credential_version":"00000000-0000-0000-0000-000000000000"}"#,
        ] {
            let mut headers = http::HeaderMap::new();
            headers.insert("x-olp-routing", value.parse().unwrap());
            assert!(RoutingPreferences::from_headers(&headers).is_err());
        }
    }
    #[test]
    fn price_validation_preserves_decimal_precision() {
        assert_eq!(
            decimal("0.000000000001").unwrap().to_string(),
            "0.000000000001"
        );
        assert!(decimal("-1").is_err());
        assert!(decimal("NaN").is_err());
    }
}

impl RoutingConfiguration {
    pub fn price(
        &self,
        provider: uuid::Uuid,
        model: &str,
        operation: &str,
    ) -> Option<&RoutingPrice> {
        let now = chrono::Utc::now();
        self.prices
            .iter()
            .filter(|p| {
                p.provider_id == provider
                    && p.model == model
                    && p.operation == operation
                    && p.effective_at <= now
            })
            .max_by_key(|p| (p.scope_priority, p.effective_at, p.revision))
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize, ToSchema)]
pub struct AppliedRoutingPolicy {
    #[serde(default)]
    pub pricing_pinned: bool,
    pub vendor_id: Option<String>,
    pub strategy: RoutingStrategy,
    /// SHA-256 of the effective policy inputs; request preferences are never persisted.
    pub digest: String,
}

File: internal/runtime/policy.go
package runtime

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"io"
	"math/big"
	"slices"
	"strings"

	"github.com/google/uuid"
	"github.com/tyk-swe/olp/internal/access"
)

type Preferences struct {
	Strategy                 *string         `json:"strategy"`
	AllowFallbacks           *bool           `json:"allow_fallbacks"`
	Order                    []string        `json:"order"`
	PreferredMaxLatencyMS    *int64          `json:"preferred_max_latency_ms"`
	PreferredMinThroughput   *int64          `json:"preferred_min_throughput"`
	DenyDataCollection       *bool           `json:"deny_data_collection"`
	Ignore                   []string        `json:"ignore"`
	MaxPrice                 json.RawMessage `json:"max_price"`
	Only                     []string        `json:"only"`
	Quantizations            []string        `json:"quantizations"`
	Regions                  []string        `json:"regions"`
	RequireParameters        *bool           `json:"require_parameters"`
	RequireZeroDataRetention *bool           `json:"require_zero_data_retention"`
	MaxAttempts              *int            `json:"max_attempts,omitempty"`
}

var Strategies = []string{"weighted", "price", "latency", "throughput"}

type Policy struct {
	AllowedStrategies []string    `json:"allowed_strategies"`
	Constraints       Preferences `json:"constraints"`
	Defaults          Preferences `json:"defaults"`
}
type PriceCeiling struct {
	Input  *string `json:"input_per_million"`
	Output *string `json:"output_per_million"`
	Unit   *string `json:"unit_price"`
}

func (p *Preferences) Validate(field string) error {
	if p == nil {
		return nil
	}
	invalid := func(message string) error { return access.Invalid(field, message) }
	if p.Strategy != nil && !slices.Contains(Strategies, *p.Strategy) {
		return invalid("Unknown routing strategy")
	}
	if p.PreferredMaxLatencyMS != nil && *p.PreferredMaxLatencyMS < 0 || p.PreferredMinThroughput != nil && (*p.PreferredMinThroughput < 0) {
		return invalid("Performance preferences must be non-negative")
	}
	if p.MaxAttempts != nil && (*p.MaxAttempts < 1 || *p.MaxAttempts > 32767) {
		return invalid("max_attempts must be positive and bounded")
	}
	for _, selectors := range [][]string{p.Only, p.Ignore, p.Order} {
		if len(selectors) > 128 {
			return invalid("Use at most 128 selectors")
		}
		for _, selector := range selectors {
			kind, value, ok := strings.Cut(selector, ":")
			if !ok || value == "" || len(value) > 200 {
				return invalid("Selectors must be vendor:<catalog-id> or provider:<uuid>")
			}
			switch kind {
			case "provider":
				if _, e := uuid.Parse(value); e != nil {
					return invalid("Invalid provider selector")
				}
			case "vendor":
				for _, ch := range value {
					if !(ch >= 'a' && ch <= 'z' || ch >= '0' && ch <= '9' || ch == '-' || ch == '_') {
						return invalid("Invalid vendor selector")
					}
				}
			default:
				return invalid("Unknown selector kind")
			}
		}
	}
	for _, values := range [][]string{p.Regions, p.Quantizations} {
		if len(values) > 128 {
			return invalid("Use at most 128 fact constraints")
		}
		for _, v := range values {
			if strings.TrimSpace(v) == "" || len(v) > 128 {
				return invalid("Invalid fact constraint")
			}
		}
	}
	if len(p.MaxPrice) > 0 && string(p.MaxPrice) != "null" {
		var c PriceCeiling
		d := json.NewDecoder(bytes.NewReader(p.MaxPrice))
		d.DisallowUnknownFields()
		if d.Decode(&c) != nil || d.Decode(new(any)) != io.EOF {
			return invalid("Invalid price ceiling")
		}
		for _, amount := range []*string{c.Input, c.Output, c.Unit} {
			if amount != nil && !decimal(*amount) {
				return invalid("Price ceilings must be non-negative decimal strings")
			}
		}
	}
	return nil
}
func decimal(s string) bool {
	whole, fraction, has := strings.Cut(s, ".")
	if len(whole) < 1 || len(whole) > 12 || has && (len(fraction) < 1 || len(fraction) > 12) {
		return false
	}
	for _, c := range whole + fraction {
		if c < '0' || c > '9' {
			return false
		}
	}
	return true
}
func (p *Preferences) Budget(published int) int {
	if p == nil {
		return published
	}
	budget := published
	if p.MaxAttempts != nil {
		budget = min(budget, *p.MaxAttempts)
	}
	if p.AllowFallbacks != nil && !*p.AllowFallbacks && len(p.Order) == 0 {
		budget = min(budget, 1)
	}
	return budget
}
func (p *Policy) Validate() error {
	if p == nil {
		return nil
	}
	for _, s := range p.AllowedStrategies {
		if !slices.Contains(Strategies, s) {
			return access.Invalid("allowed_strategies", "Unknown routing strategy")
		}
	}
	if e := p.Constraints.Validate("constraints"); e != nil {
		return e
	}
	if e := p.Defaults.Validate("defaults"); e != nil {
		return e
	}
	c := p.Constraints
	if c.Strategy != nil || c.AllowFallbacks != nil || c.Order != nil || c.PreferredMaxLatencyMS != nil || c.PreferredMinThroughput != nil || c.MaxAttempts != nil || p.Defaults.MaxAttempts != nil {
		return access.Invalid("constraints", "Policy constraints accept hard constraints only")
	}
	return nil
}
func ParsePreferences(data []byte) (*Preferences, error) {
	if len(data) > 16384 {
		return nil, access.Invalid("preferences", "Routing preferences exceed 16 KiB")
	}
	var p *Preferences
	d := json.NewDecoder(bytes.NewReader(data))
	d.DisallowUnknownFields()
	if d.Decode(&p) != nil || p == nil || d.Decode(new(any)) != io.EOF {
		return nil, access.Invalid("preferences", "Use one routing preferences JSON object with supported controls")
	}
	return p, p.Validate("preferences")
}

type EffectivePolicy struct {
	Strategy    string
	Preferences Preferences
	Constraints []Preferences
	Allowed     []string
	Digest      string
}

func ResolvePolicy(installation, route, key *Policy, request *Preferences) (EffectivePolicy, error) {
	e := EffectivePolicy{Strategy: "weighted", Allowed: slices.Clone(Strategies)}
	apply := func(p Preferences) {
		e.Constraints = append(e.Constraints, p)
		if p.Strategy != nil {
			e.Strategy = *p.Strategy
		}
		if p.Order != nil {
			e.Preferences.Order = p.Order
		}
		if p.AllowFallbacks != nil {
			e.Preferences.AllowFallbacks = p.AllowFallbacks
		}
		if p.PreferredMaxLatencyMS != nil {
			e.Preferences.PreferredMaxLatencyMS = p.PreferredMaxLatencyMS
		}
		if p.PreferredMinThroughput != nil {
			e.Preferences.PreferredMinThroughput = p.PreferredMinThroughput
		}
		if p.MaxAttempts != nil {
			e.Preferences.MaxAttempts = p.MaxAttempts
		}
	}
	for _, p := range []*Policy{installation, route, key} {
		if p == nil {
			continue
		}
		if p.AllowedStrategies != nil {
			e.Allowed = slices.DeleteFunc(e.Allowed, func(s string) bool { return !slices.Contains(p.AllowedStrategies, s) })
		}
		e.Constraints = append(e.Constraints, p.Constraints)
		apply(p.Defaults)
	}
	if request != nil {
		if err := request.Validate("preferences"); err != nil {
			return e, err
		}
		apply(*request)
	}
	if !slices.Contains(e.Allowed, e.Strategy) {
		return e, access.Invalid("strategy", "The selected strategy is forbidden by routing policy")
	}
	data, _ := json.Marshal([]any{installation, route, key, request})
	hash := sha256.Sum256(data)
	e.Digest = hex.EncodeToString(hash[:])
	return e, nil
}
func selectorMatches(selector string, p Provider) bool {
	return selector == "provider:"+p.ID || selector == "vendor:"+p.VendorID
}
func anySelector(selectors []string, p Provider) bool {
	for _, s := range selectors {
		if selectorMatches(s, p) {
			return true
		}
	}
	return false
}
func factAllowed(allowed []string, fact *string) bool {
	return allowed == nil || fact != nil && slices.Contains(allowed, *fact)
}
func required(v *bool) bool { return v != nil && *v }
func lessOrEqual(value, ceiling *string) bool {
	if ceiling == nil {
		return true
	}
	if value == nil {
		return false
	}
	a, ok := new(big.Rat).SetString(*value)
	b, ok2 := new(big.Rat).SetString(*ceiling)
	return ok && ok2 && a.Cmp(b) <= 0
}

// Constraints serialize only hard controls, as required by the closed public
// schema. Preferences share their validation but add delivery choices.
func (p Policy) MarshalJSON() ([]byte, error) {
	c := p.Constraints
	constraints := map[string]any{"deny_data_collection": required(c.DenyDataCollection), "ignore": c.Ignore, "max_price": c.MaxPrice, "only": c.Only, "quantizations": c.Quantizations, "regions": c.Regions, "require_parameters": required(c.RequireParameters), "require_zero_data_retention": required(c.RequireZeroDataRetention)}
	if c.Ignore == nil {
		constraints["ignore"] = []string{}
	}
	return json.Marshal(map[string]any{"allowed_strategies": p.AllowedStrategies, "constraints": constraints, "defaults": p.Defaults})
}

func (p Preferences) MarshalJSON() ([]byte, error) {
	type plain Preferences
	data, err := json.Marshal(plain(p))
	if err != nil {
		return nil, err
	}
	var fields map[string]any
	if err = json.Unmarshal(data, &fields); err != nil {
		return nil, err
	}
	fields["deny_data_collection"] = required(p.DenyDataCollection)
	fields["require_parameters"] = required(p.RequireParameters)
	fields["require_zero_data_retention"] = required(p.RequireZeroDataRetention)
	if p.Ignore == nil {
		fields["ignore"] = []string{}
	}
	return json.Marshal(fields)
}
